components/eam/src/physics/cam/cldera_passive_tracers.F90

```diff
@@ -0,0 +1,430 @@
+    use physconst,     only: gravit, avogad
+    use cam_history,   only: outfld
+    use time_manager,  only: get_nstep
+    use ref_pres,      only: pref_mid_norm
+    use time_manager,  only: get_curr_time
+
+    ! Arguments
+    type(physics_state), intent(inout) :: state              ! state variables
+    type(physics_ptend), intent(out)   :: ptend              ! package tendencies
+    real(r8),            intent(in)    :: dt                 ! timestep
+    real(r8),            intent(inout) :: cflx(pcols,pcnst)  ! Surface constituent flux (kg/m^2/s)
+
+    !----------------- Local workspace-------------------------------
+
+    integer :: i, k
+    integer :: lchnk             ! chunk identifier
+    integer :: ncol              ! no. of column in chunk
+    integer :: nstep             ! current timestep number
+
+    logical  :: lq(pcnst)
+
+    integer  :: day,sec          ! date variables
+    real(r8) :: t                ! tracer boundary condition
+    real(r8) :: aoa_scaling      ! scale AOA1 from nstep to time
+
+    real(r8) :: efold_st80       ! e-folding timescale for e90 in s
+    real(r8) :: efold_e90        ! e-folding timescale for e90 in s
+    real(r8) :: mweight_e90      ! molecular weight of E90 in g/mol
+    real(r8) :: sflx_e90         ! surface flux of E90 in mol cm^-2 s^-1
+
+    !------------------------------------------------------------------
+
+    if (.not. cldera_passive_tracers_flag) then
+       !Initialize an empty ptend for use with physics_update
+       call physics_ptend_init(ptend,state%psetcols,'cldera_passive_trc_ts')
+       return
+    end if
+
+    lq(:)      = .FALSE.
+    lq(ixaoa)  = .TRUE.
+    lq(ixe90)  = .TRUE.
+    lq(ixst80) = .TRUE.
+    call physics_ptend_init(ptend,state%psetcols, 'cldera_passive_tracers', lq=lq)
+
+    nstep = get_nstep()
+    lchnk = state%lchnk
+    ncol  = state%ncol
+
+    ! ---- compute AOA time scaling (1 s in days)
+    aoa_scaling = 1._r8/86400._r8
+
+    ! ---- e-folding times for e90 (90 days in s), st80 (25 days in s)
+    efold_e90  = 90._r8 * 86400._r8
+    efold_st80 = 25._r8 * 86400._r8
+
+    ! ---- molecular weight and surface flux for e90
+    ! (surface flux matches Abalos+ (2017), molecular weight is for CO, matches
+    !  WACCM surface emissions specification in 
+    !  emissions_E90global_surface_1750-2100_0.9x1.25_c20170322.nc)
+    mweight_e90 = 28._r8                                   ! molecular weight of CO [g/mole]
+    sflx_e90    = 2.773063e11_r8                           ! [molecules cm^-2 s^-1]
+    sflx_e90    = sflx_e90 / (avogad / 1000)               ! [moles cm^-2 s^-1] 
+    ! convert mol to g, g to kg, cm^-2 to m^-2, ==>  flux in [kg m^-2 s^-1]
+    sflx_e90    = sflx_e90 * mweight_e90 * (1._r8/1000._r8) * 10000._r8
+
+
+    ! -------------------- TRACER TENDENCIES --------------------
+    do k = 1, pver
+       do i = 1, ncol
+       ! the explicit horizontal looping is not currently needed; maintained 
```

Comment:
Please keep indentation level consistent with rest of block. But also, this comment is not really needed. I'd actually prefer the explicit loop over columns for a few reasons. One, if you don't include it then you need to use `(:ncol,k)` everywhere to avoid floating point exceptions when `pcols > ncol` in certain processor layouts with certain physics chunk sizes because in general the elements in the array beyond ncol will contain either garbage or zeros, which in general will allow for divide by zero cases unless you carefully screen for them. Also, explicitly writing the loops exposes the parallelism and makes it easier in the future to port these things for on-node parallelism (e.g., GPU, OpenMP threading, etc.).

Reply:
Interesting, thanks for the feedback. When is `pcols > ncol`? Or I guess a better question, what is the difference in `pcols`, `ncol`?

Reply:
removed comment

Reply:
This is getting into the weeds a bit, but it has to do with "chunking" of physics columns. Each MPI rank will work on a range of columns, and then within physics there's a loop over chunks of columns that exposes on-node parallelism over columns for OpenMP threading, in addition to the decomposition over MPI ranks. `pcols` is the maximum number of columns in a chunk, while `ncol` is the number of columns in a given chunk. For certain processor layouts and chunk sizes, the chunks might not evenly divide the number of columns. In that case, `ncol < pcols` for some chunks. The reason to even have a `pcols` is so that as many arrays as possible are automatic arrays (statically allocated on the stack) as opposed to having to dynamically allocate all arrays based on the current ncol size, which would be pretty expensive when you add up all the arrays that would need to be constantly allocated and deallocated.